Adding Parsec support in Edge Core

whygoyal opened the pull request:
1. Added `rust-layer` and `clang-layer` as LAYERDEPENDS. Also, added those layers to manifest-pelion-edge. 
2. Added patch to disable the default edge-core CMake commands to compile `parsec-se-driver`.
3. Added separate recipe for `parsec-se-driver` and added it as DEPENDS in mbed-edge-core. Before compiling edge-core, the  `libparsec-se-driver.a` is then placed in the folder where factory-configurator-client is expecting to be. 
4. By default, PARSEC_TPM_SE_SUPPORT is turned off for all platform. Parsec is only supported in factory flow, so to compile parsec client in edge-core, you have to turn on FACTORY_MODE and PARSEC_TPM_SE_SUPPORT. 
5. Upgraded mbed-edge-core to 0.16.0 and updated the patches to work with it.
6. Added python3 as depends. 

Compiles in dev environment. 
Has not been verified on RPi as it depends on Parsec service and software TPM which has not been integrated.
Has not been tested with LmP.

Commit:
Added dependencies required to build parsec client

Commit:
Added LAYERDEPENDS - rust, clang

Commit:
Added cargo-native as depends

Commit:
Added PARSEC_TPM_SE_SUPPORT build configurations to target.cmake

Also added a patch to check if cargo is installed before starting the build

Commit:
Added parsec-se-driver as DEPENDS, mbed-edge-core to 0.16.0 and updated patches

Commit:
Only require parsec-se-driver when PARSEC_TPM_SE_SUPPORT is enabled

Commit:
Copy the parsec-se-driver.a only when parsec is required

petedyerarm commented on recipes-connectivity/mbed-edge-core/files/rpi3/target.cmake:
```diff
@@ -5,11 +5,24 @@ SET (PAL_TARGET_DEVICE "Yocto_Generic")
 if (${FIRMWARE_UPDATE})
   SET (MBED_CLOUD_CLIENT_UPDATE_STORAGE ARM_UCP_LINUX_YOCTO_RPI)
 endif()
+
```
Is this change also required for imx8mmevk and uz ?

ygoyal18 replied:
Yes, its required.

costanic commented on recipes-connectivity/mbed-edge-core/files/0003-fix_libwebsockets_nondebug.manual_patch:
```diff
@@ -2,18 +2,12 @@ diff --git a/lib/libwebsockets/libwebsockets/CMakeLists.txt b/lib/libwebsockets/
 
 --- a/lib/libwebsockets/libwebsockets/CMakeLists.txt
 +++ b/lib/libwebsockets/libwebsockets/CMakeLists.txt
-@@ -1043,13 +1043,13 @@
```
At quick glance, I don't see why this patch file is changed in the context of this PR.  Does libwebsockets need to be compiled for parsec in some way, or was there some other change in this PR that required a change here?

ygoyal18 replied:
Edge Core 0.16.0 is required to successfully build Parsec client so that is also upgraded in this recipe. And inturn that requires us to update the patches.

costanic commented on recipes-connectivity/mbed-edge-core/files/0004-Removed-the-redundant-cmake-command.manual_patch:
```diff
@@ -1,24 +0,0 @@
-From dae10c64256f147cd5ea86013cd697a512f22e45 Mon Sep 17 00:00:00 2001
```
Same question here, was there a change somewhere else that means this patch is no longer needed, in the context of adding Parsec support?

ygoyal18 replied:
Ya, this is related to 0.16.0. This patch was upstreamed in that version.

costanic commented on recipes-connectivity/mbed-edge-core/files/0005-fota.manual_patch:
```diff
@@ -1,23 +1,25 @@
  #endif
- 
+
+ #define MBED_CLOUD_CLIENT_FOTA_LINUX_TEMP_HEADER_FILENAME MBED_CLOUD_CLIENT_FOTA_LINUX_HEADER_FILENAME ".tmp"
```
same question here.  what's the connection with Parsec support?

costanic commented on recipes-connectivity/mbed-edge-core/files/rpi3/target.cmake:
```diff
@@ -5,11 +5,24 @@ SET (PAL_TARGET_DEVICE "Yocto_Generic")
+
+if(${PARSEC_TPM_SE_SUPPORT})
+  SET (PAL_FS_MOUNT_POINT_PRIMARY "\"/userdata/mbed/psa\"")
+  SET (PAL_FS_MOUNT_POINT_SECONDARY "\"/userdata/mbed/psa\"")
```
Why does this path need to change?  This will cause other scripts to break that have hard-coded the mcc_config path, such as https://github.com/PelionIoT/meta-mbed-edge/blob/87802c415b59d359c5928ef8a2d503df28d4bde2/recipes-connectivity/mbed-edge-core/files/edge-core.init#L14.  I'm not sure but there might be others, like the firmware update scripts.

ygoyal18 replied:
Thanks @costanic. This is a good question, I will check if we can rename the `psa` folder or not.. This is the default folder generated by FCC (in mbed-cloud-client).

ygoyal18 replied:
FYI @moshe-shahar

Commit:
Updated edge-core to 0.16.1

Commit:
Added recipe to separately compile parsec-se-driver (#31)

* Added recipe to separately compile parsec-se-driver

* Updated the parsec-se-driver recipe to follow standard Yocto approach

* using yocto macros `libdir` to refer to /usr/lib

Commit:
Export proxy values for pip3 and python3

JanneKiiskila commented on recipes-connectivity/mbed-edge-core/files/0003-fix_libwebsockets_nondebug.manual_patch:
```diff
@@ -2,18 +2,12 @@ diff --git a/lib/libwebsockets/libwebsockets/CMakeLists.txt b/lib/libwebsockets/
+@@ -1237,7 +1237,7 @@ if (CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX OR COMPILER_IS_CLANG)
      if (CMAKE_BUILD_TYPE MATCHES "DEBUG")
- 		set(CMAKE_C_FLAGS "-O0 ${CMAKE_C_FLAGS}")
+                set(CMAKE_C_FLAGS "-O0 ${CMAKE_C_FLAGS}")
```
tabs/spaces issue?

ygoyal18 replied:
Not sure.. I just regenerated the patch without touching those lines. Could be that it got changed in 3.1-arm version of libwebsocket.

Commit:
Define branch when sourcing parsec-se-driver

JanneKiiskila commented on recipes-connectivity/parsec-se-driver/parsec-se-driver_0.4.0.bb:
```diff
@@ -4,7 +4,7 @@ SUMMARY = "Parsec Secure Element Driver"
 LICENSE = "Apache-2.0"
 
-SRC_URI = "git://github.com/parallaxsecond/parsec-se-driver.git;protocol=https \
+SRC_URI = "git://github.com/parallaxsecond/parsec-se-driver.git;protocol=https;branch=main \
```
Should it not pick up the tag 0.4.0? As the recipe name implies? It might be dangerous to update the SE driver to a newer version, as the APIs seem to change a bit (at least towards MbedTLS). SE Driver 0.5.0 will have support for Mbed TLS 2.25.0.

ygoyal18 replied:
@JanneKiiskila The build is actually failing with this error if we don't specify the branch 
```
ERROR: parsec-se-driver-0.3.1-r0 do_fetch: Fetcher failure: Unable to find revision baccf6682818edd6ae182e2f3144f1b0afb47c92 in branch master even from upstream
```
looks like there is no longer `master` branch and its been replaced with `main`. Not sure why Yocto is not able to fetch the 0.4.0 without specifying the branch.

JanneKiiskila replied:
Hmmh... Interesting. Yes, the "politically correct" naming thing has real engineering impact. It seems bitbake defaults to accepting hashes from master, but not from main. The hash exists, it just doesn't exist in "master branch" anymore.

JanneKiiskila replied:
https://github.com/parallaxsecond/parsec-se-driver/commit/baccf6682818edd6ae182e2f3144f1b0afb47c92 - that commit ID is still there, it hasn't gone anywhere. It's just not on `master` branch, but in `main` instread.

Commit:
Update the rpi target.cmake (#37)

* Update the rpi target.cmake

* Applied suggested change

Commit:
Updated Xilinx board target.cmake to enable Parsec (#36)

* Updated Xilinx board target.cmake to enable Parsec

* Applied suggested change

Commit:
Add parsec support to imx8 (#35)

* Initial version of target.cmake changes

* Initial version of target.cmake changes

* Fix PSA storage location

* Updated the psa storage to /userdata/mbed/mcc_config and also imx8's target.cmake

Co-authored-by: petedyerarm <pete.dyer@arm.com>
Co-authored-by: Moshe Shahar <moshe.shahar@arm.com>

Commit:
Follow `dash` rules as The scripts are executed by /bin/sh, which may not be a bash shell but might be something such as dash. You should not use Bash-specific script (bashisms)

https://www.yoctoproject.org/docs/2.3.4/bitbake-user-manual/bitbake-user-manual.html#shell-functions
